sklearn/feature_selection/mutual_info.py

```diff
@@ -0,0 +1,230 @@
+        features. Note, that diagonal elements are not computed and contain
+        some random numbers.
+
+    References
+    ----------
+    .. [1] H. Peng, F. Long, and C. Ding, "Feature selection based on mutual
+           information: criteria of max-dependency, max-relevance, and
+           min-redundancy", IEEE Transactions on Pattern Analysis and Machine
+           Intelligence, Vol. 27, No. 8, pp. 1226-1238, 2005.
+    .. [2] A. Kraskov, H. Stogbauer and P. Grassberger, "Estimating mutual
+           information". Phys. Rev. E 69, 2004.
+    .. [3] B. C. Ross "Mutual Information between Discrete and Continuous
+           Data Sets". PLoS ONE 9(2), 2014.
+    """
+    def __init__(self, n_features_to_select=0.5, categorical_features=False,
+                 categorical_target=False, n_neighbors=3):
+        self.n_features_to_select = n_features_to_select
+        self.categorical_features = categorical_features
+        self.categorical_target = categorical_target
+        self.n_neighbors = n_neighbors
+
+    def _get_support_mask(self):
+        return self.support_
+
+    def fit(self, X, y):
+        """Find the best features according to mutual information criterion.
+
+        Parameters
+        ----------
+        X : array_like or sparse matrix, shape (n_samples, n_features)
+            Feature matrix.
+        y : array_like, shape (n_samples,)
+            Target vector.
+
+        Returns
+        -------
+        self
+        """
+        X, y = check_X_y(X, y, accept_sparse='csc',
+                         y_numeric=not self.categorical_target)
+
+        n_features = X.shape[1]
+
+        if self.n_features_to_select >= 1:
+            n_features_to_select = int(self.n_features_to_select)
+        elif 0 < self.n_features_to_select < 1:
+            n_features_to_select = max(
+                1, int(self.n_features_to_select * n_features))
+        else:
+            raise ValueError("`n_features_to_select` must be positive.")
+
+        if isinstance(self.categorical_features, bool):
+            categorical_features = np.empty(n_features, dtype=bool)
+            categorical_features.fill(self.categorical_features)
+        else:
+            categorical_features = self.categorical_features
+
+        relevance = np.empty(n_features)
+        redundancy = np.empty((n_features, n_features))
+
+        for i in range(n_features):
+            if issparse(X):
+                xi = X[:, i].toarray().ravel()
+            else:
+                xi = X[:, i]
+
+            relevance[i] = _compute_mi(
+                xi, y, categorical_features[i],
+                self.categorical_target, self.n_neighbors)
+
+            for j in range(i + 1, n_features):
+                if issparse(X):
+                    xj = X[:, j].toarray().ravel()
+                else:
+                    xj = X[:, j]
+
+                redundancy[i, j] = redundancy[j, i] = _compute_mi(
```

Comment:
I thought about that. If we choose to do it that way then should we provide `redundancy_` attribute (only partially computed)?

Related thought: is it better to actually select `n_features_to_select` features or do a full feature ranking? 

In the first case it looks logical (and more efficient) to compute MI only between required pairs of variables (as you suggested). Then probably we shouldn't provide any attribute except for `support_`.

Another approach is to do full feature ranking and return MI between each pair of variables, and variable-target MI. 

Now it is something in between (doesn't seem good). So which approach you think is better?